### DISCUSSION TITLE
Halide::Runtime::Buffer does not use halide_malloc and halide_free.

### DISCUSSION THREAD

**Discussion by mcourteaux** [halide/Halide#6171] [Q&A]
https://github.com/halide/Halide/blob/a914574d504bc137a9e23545bbbf5d5504f1c485/src/runtime/HalideBuffer.h#L743

The documentation of the class says that the default allocator functions are malloc and free, whereas the HalideRuntime.h specifies halide_malloc and halide_free for memory management.

Additionally, Halide::Runtime::Buffer does not allow for creating buffers with other allocators. You always go through an allocate() call using the default malloc/free, which is wasteful if you want to use a custom allocator afterwards.

**Comment by steven-johnson** [comment]
> You always go through an allocate() call using the default malloc/free

It's awkward, but you should be able to avoid this by calling the ctor with a NULL data ptr, then an explicit call to allocate(), eg

```
    std::vector<halide_dimension_t> dims = ...;
    HalideBuffer<uint8_t> new_buffer(nullptr, (int)dims.size(), dims.data());
    new_buffer.allocate(my_allocate_fn, my_deallocate_fn);
```

**Reply by mcourteaux** [reply]
But what's the point of the Runtime::Buffer not respecting the choice of allocator in the Runtime? The workaround you present is not super bad, but in general I'm thinking that the situation might improve when it respects the allocator. This would be super useful to integrate Tracy for memory tracking.

**Comment by mcourteaux** [comment]
@alexreinking Why was this closed? I'm again running in this problem. Why does it not respect the custom allocator set in the Runtime?

**Comment by steven-johnson** [comment]
Tagging @abadams

**Comment by abadams** [comment]
Halide::Runtime::Buffer is used in contexts where there is no runtime linked. Specifically, it's what backs Halide::Buffer, which is what is used when using Halide as a JIT compiler.

**Reply by mcourteaux** [reply]
Ah, some sort of separation of concerns going on here. Can I make a PR where we can set a default custom allocator for this, similar to how it's done in `HalideRuntime.h`, along with some documentation clarifications regarding this?

I'm instrumenting my application with Tracy, and would like to track memory allocations/frees. Many of them are hiding in `Halide::Runtime::Buffer`. For now, I just manually edited the `HalideRuntime.h` header file to do the necessary Tracy work, but this is not a durable solution.

**Reply by abadams** [reply]
You mean something like a protected static member of Halide::Runtime::Buffer which gives the default allocator to use and a method to set it? Seems reasonable to me.

**Reply by mcourteaux** [reply]
Yes! Oki, I'll look at it tomorrow! Thanks for the quick reply again! :smile:

**Reply by mcourteaux** [reply]
I am a bit puzzled with where to put the static variables that will hold the default function pointers. As it's C++ you can't put them in a header. But I can't find any appropriate cpp file that could host these two variables.

**Reply by abadams** [reply]
C++17 lets you declare static members in headers as having inline linkage:
```
class Foo {
  static inline int bar;
};
```

I haven't used them before but in a quick test they seem to work. If I include that class in two distinct translation units, and then link them, I get the same address for Foo::bar in both translation units.

**Reply by mcourteaux** [reply]
PR #8132.
